File: leetcode/serialize_deseralize_binary_tree.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:
    def serialize(self, root):
        queue = collections.deque([root])
        # initalize index 0 as # and only use from index 1
        result = ['#']
        # tree bfs serialization
        while queue:
            node = queue.popleft()
            if node:
                queue.append(node.left)
                queue.append(node.right)
                result.append(str(node.val))
            else:
                result.append('#')
        return ' '.join(result)

    def deserialize(self, data):
        # exception
        if data == '# #': # front is # and end is # too
            return None
        
        nodes = data.split()
        root = TreeNode(nodes[1]) # starting from index 1
        queue = collections.deque([root]) # initalize queue with index 1
        index = 2
        # using fast runner strategy to check the child node result in advance
        while queue:
            node = queue.popleft()
            if (nodes[index] != '#'):
                node.left = TreeNode(int(nodes[index]))
                queue.append(node.left)
            index += 1
            if (nodes[index] != '#'):
                node.right = TreeNode(int(nodes[index]))
                queue.append(node.right)
            index += 1
        return root
        
# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: leetcode/serialize_deseralize_binary_tree.js
function TreeNode(val) {
    this.val = val;
    this.left = this.right = null;
};

let serialize = function(root) {
    if (root != null && root.val === undefined) {
        return "#";
    };
    let result = ['#']
    let queue = [root];
    while (queue.length > 0) {
        let node = queue.shift();
        if (node !== null) {
            queue.push(node.left);
            queue.push(node.right);
            result.push(node.val.toString());
        } else {
            result.push('#');
        };
    };
    return result.toString();
};

let deserialize = function(temp) {
    if (temp === '# #') {
        return [];
    };
    if (temp === '#') {
        return [];
    };
    let data = temp.replace(/,/g, "");
    let nodes = data.split("");
    let root = new TreeNode(Number.parseInt(nodes[1]));
    let queue = [root];
    let index = 2;
    while (queue.length > 0) {
        let node = queue.shift();
        if (nodes[index] !== '#') {
            node.left = new TreeNode(Number.parseInt(nodes[index]));
            queue.push(node.left);
        };
        index += 1;
        if (nodes[index] !== '#') {
            node.right = new TreeNode(Number.parseInt(nodes[index]));
            queue.push(node.right);
        };
        index += 1;
    };
    return root;
};

let one = new TreeNode(1);
let two = new TreeNode(2);
let three = new TreeNode(3);
let nothing = new TreeNode(null);
let four = new TreeNode(4);
let five = new TreeNode(5);
one.left = two
one.right = three
three.left = four
three.right = five
console.log(serialize(deserialize(serialize(one))));

let nothing_two = new TreeNode();
console.log(serialize(deserialize(serialize(nothing_two))));